### PR TITLE
chore: Refactor install.ps1

### PR DESCRIPTION
**Pull request by bradenhilton**
I've been wanting to do this for a long time but always seemed to find something else to do instead.

Changes:

- Use PowerShell's [Approved Verbs](https://learn.microsoft.com/en-us/powershell/scripting/developer/cmdlet/approved-verbs-for-windows-powershell-commands?view=powershell-5.1)
- Use PowerShell's built-in output streams
- Remove unused functions
- General style changes to align more with the [unofficial style guide](https://poshcode.gitbooks.io/powershell-practice-and-style)
- ~Use GitHub's REST API (I am concerned this may cause issues with actions but I'm not sure. If this is preferred I can also make this change in `install.sh.tmpl`)~
- Removed almost 100 lines

`get_goos` handled FreeBSD, but I'm struggling to even find any information about PowerShell being installable on FreeBSD. https://github.com/PowerShell/PowerShell?tab=readme-ov-file#get-powershell makes no mention of FreeBSD, nor does https://github.com/dotnet/core/blob/main/release-notes/8.0/supported-os.md.

```console
PowerShell 7.4.1

~
❯ [System.Runtime.InteropServices.RuntimeInformation]::FrameworkDescription
.NET 8.0.1
```

https://ports.freebsd.org/cgi/ports.cgi?query=powershell&stype=all&sektion=all
https://ports.freebsd.org/cgi/ports.cgi?query=pwsh&stype=all&sektion=all

TODO:

- [X] Verify that the invocation remains the same when downloading and executing the script over the network
- [X] Test on Linux and macOS

There's no rush for this. I want to be absolutely sure there are no regressions before merging.

### CODE REVIEW

**Comment by bradenhilton** [comment]
I'll test locally on my MBP tomorrow.

The `trap` statement is causing the issue at the moment, once it's fixed (or I switch back to `try..catch..finally`) I can move on to whatever error is causing it.

**Comment by twpayne** [comment]
> The `trap` statement is causing the issue at the moment, once it's fixed (or I switch back to `try..catch..finally`) I can move on to whatever error is causing it.

I think the problem is that the temporary directory is still the process's working directory when the `trap` statement is invoked. I would suggest trying to avoid changing the working directory at all (i.e. make the script work without the call to `Push-Location $tempDir`).

**Comment by bradenhilton** [comment]
The error is caused by the `tar` argument `-file`, which I could have sworn I saw in the man page. I just checked again and confirmed that it's definitely `--file`. Strange.

**Comment by bradenhilton** [comment]
> I would suggest trying to avoid changing the working directory at all (i.e. make the script work without the call to `Push-Location $tempDir`).

I think I'll do this anyway. I generally dislike pushing/popping directories.

I also don't think most of the comments are particularly useful any more.

**Comment by bradenhilton** [comment]
`format-yaml.py` doesn't explicitly set line endings when writing the file. Separate fix coming up.

**Comment by bradenhilton** [comment]
I'm not sure why the trailing newline is causing an issue in [`main/check`](https://github.com/twpayne/chezmoi/actions/runs/7934435374/job/21665421235?pr=3569). If I run `git restore -s upstream/master ./.github/workflows/main.yml` it shows a trailing newline. I wonder if it's adding a second one somehow.

I re-applied my changes over the restored workflow file, then ran it through the updated `format-yaml.py` from https://github.com/twpayne/chezmoi/pull/3575.

I'm also not sure why it added newlines at lines 400 and 503.

**Comment by twpayne** [comment]
Note that it's [editorconfig-checker](https://github.com/editorconfig-checker/editorconfig-checker) that's complaining about the whitespace.

I'd suggest manually reverting the unnecessary whitespace changes. Or, alternatively, feel free to merge this as-is and I'll investigate editorconfig-checker and `format-yaml.py` separately.

**Comment by twpayne** [comment]
The changes to the behavior of `format-yaml.py` are likely due to changes in [the `ruamel.yaml` package that it uses](https://yaml.readthedocs.io/en/latest/).

**Comment by twpayne** [comment]
It's also `ruamel.yaml` that adds the trailing whitespace. 😢

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -335,12 +347,12 @@ jobs:
...
       run: |
         if (Test-Path -Path bin/chezmoi.exe) { Remove-Item -Force bin/chezmoi.exe }
-        (irm -useb https://get.chezmoi.io/ps1) | powershell -c -
+        iex "&{$(irm 'https://get.chezmoi.io/ps1')}"
         bin/chezmoi.exe --version
     - name: test-install-iwr
       run: |
         if (Test-Path -Path bin/chezmoi.exe) { Remove-Item -Force bin/chezmoi.exe }
-        (iwr -useb https://get.chezmoi.io/ps1).ToString() | powershell -c -
+        iex "&{$(irm 'https://get.chezmoi.io/ps1')}"
```
Comment (bradenhilton) [inline .github/workflows/main.yml:355]: Acknowledging the typo here

**Comment by bradenhilton** [comment]
@twpayne Would you be willing to test both the local `install.ps1` and the `iex "&{$(irm 'URL')} ARGS..."` method with either the raw link to `install.ps1` or [this gist](https://gist.githubusercontent.com/bradenhilton/706525977a09fa98a3b6d6c0552aae7b/raw/bf69d265fe25a966447cc2cd0a61e77094e6aade/install.ps1) on Linux?

If you aren't familiar, you should be able to run

```console
❯ ./path/to/install.ps1
```

and

```console
❯ iex "&{$(irm 'URL')} ARGS..."
```

after launching PowerShell with `pwsh`.

I am particularly interested in any potential issues `Get-LibC` might have, as it's the only thing I haven't tested yet, I just tried to model it after `install.sh`.

Should I also add the `iex "&{$(irm 'URL')} ARGS..."` method to `test-macos` and `test-ubuntu`?

**Comment by twpayne** [comment]
On Ubuntu (glibc-based) it works perfectly.

On Alpine Linux (musl-based) the libc detection fails. It looks like an error should be ignored and is not. Here's how I tested:

```console
$ docker pull alpine:latest
$ docker run -it alpine:latest /bin/sh
# apk add powershell
...
# pwsh
PowerShell 7.3.10
PS /> 
PS /> iex "&{$(irm 'URL')}"
...
musl libc (x86_64)
Version 1.2.4_git20230717
Dynamic Program Loader
Usage: /lib/ld-musl-x86_64.so.1 [options] [--] pathname
getconf: GNU_LIBC_VERSION: unknown variable
```

Note that there's a bug in Alpine Linux/PowerShell that causes it to echo the whole line typed so far each time you type a character. However, the command does run.

For the `URL` I used https://raw.githubusercontent.com/twpayne/chezmoi/30cbffff9b7c0c9ad0081541c750e12520cea713/assets/scripts/install.ps1 which gave the raw content of `install.ps1` in this PR at the time of writing (in https://github.com/twpayne/chezmoi/pull/3569/files, click the three dots to the right of the filename, chose "View file", and then "Raw").

**Comment by twpayne** [comment]
Ah, a small bug: running `install.ps1` adds all files in the tarball (`README.md`, `LICENSE`, and `completions/*` to the current directory. It should probably only add the `chezmoi` binary.

**Comment by bradenhilton** [comment]
The new `test-install.ps1-irm`s that I added to `test-macos` and `test-ubuntu` are currently still using the old version of the script.

I'll try temporarily permalinking to the raw file using `GITHUB_SHA` (if I can figure out the correct syntax...)

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -135,6 +135,18 @@ jobs:
...
+      shell: pwsh
+      run: |
+        if (Test-Path -Path bin/chezmoi.exe) { Remove-Item -Force bin/chezmoi.exe }
+        assets/scripts/install.ps1
+        bin/chezmoi --version
+    - name: test-install.ps1-irm
+      shell: pwsh
+      run: |
+        if (Test-Path -Path bin/chezmoi.exe) { Remove-Item -Force bin/chezmoi.exe }
+        iex "&{$(irm 'https://raw.githubusercontent.com/twpayne/chezmoi/${{ env.GITHUB_SHA }}/assets/scripts.install.ps1')}"
```
Comment (twpayne) [inline .github/workflows/main.yml:148]: Should this be `/assets/scripts/install.ps1` instead of `/assets/scripts.install.ps1`? (replace `.` with `/`)?

Comment (bradenhilton) [reply .github/workflows/main.yml:148]: Yes

**Comment by bradenhilton** [comment]
If `${{ github.event.pull_request.head.sha }}` doesn't work I might try capturing the output of `git rev-parse HEAD` in a variable and using that.

**Comment by bradenhilton** [comment]
Seems like `${{ github.event.pull_request.head.sha }}` works, but now the script is failing to install chezmoi (specifically, failing somewhere in the middle, as it at least prints `found version 2.46.1 for latest/.../amd64`), despite the same script working when run locally.

I will fill the script with debug logs tomorrow to see if I can figure it out.

**Comment by twpayne** [comment]
Thank you again for your work here. Sleep well and we can debug this tomorrow :)

**Comment by bradenhilton** [comment]
Hopefully removing `.exe` from the macOS/Ubuntu steps is one of the last remaining issues.

**Comment by bradenhilton** [comment]
@twpayne I haven't merged this yet because the API request fails for me on alpine:

```console
PS /> .\install.ps1 -d
DEBUG: BinDir: /bin
DEBUG: Tag: latest
DEBUG: ChezmoiArgs:
DEBUG: tempDir: /tmp/90a8e3bf-3d1d-4825-8468-340f3f3cd506
DEBUG: goOS: linux
DEBUG: goArch: amd64
DEBUG: endpoint: https://api.github.com/repos/twpayne/chezmoi/releases/latest
DEBUG: removing /tmp/90a8e3bf-3d1d-4825-8468-340f3f3cd506
Invoke-RestMethod: /install.ps1:100
Line |
 100 |      Invoke-RestMethod -Uri $endpoint -Headers $headers
     |      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Resource temporarily unavailable
```

It has definitely worked previously, and I haven't changed any critical code related to the request, so I'm not entirely sure what the issue is. I tried assigning the result to a variable and inspecting it, but it doesn't even get that far.

I was also running into rate limit issues while debugging yesterday. I was considering checking for environment variables to use for authentication but the API docs don't explicitly recommend storing tokens in environment variables.

I think for now I will revert to the same method used in `install.sh` and see how that goes.

**Comment by twpayne** [comment]
The integration tests are occasionally flaky as they rely on a bunch of third-party infrastructure working, and the chance that none of them have a problem at any given moment is depressingly small. I've notice Alpine being a bit flaky recently.

Normally I re-run all failed jobs if the cause of failure is not clearly in chezmoi, and only investigate in detail if the job fails a second time.

**Comment by bradenhilton** [comment]
We don't test the install script on alpine with a job though. I was testing it manually in a fresh container. All I've been doing is

```console
❯ docker run -it alpine:latest

/ # apk update
...

/ # apk add powershell
...

/ # pwsh

PS /> iex "&{$(irm 'URL')} -d"
all requests to github fail...
```

I still can't get it to work. For now I'm going to assume the script isn't at fault.

File: assets/scripts/install.ps1 [hunk 3]
```diff
@@ -1,322 +1,264 @@
...
+Move-Item -Path $tempBinaryPath -Destination $binary -Force
+Write-Information "installed ${binary}"
 
-try {
-    Invoke-Expression ("Install-Chezmoi " + $params)
-} catch {
-    Write-Host "An error occurred while installing: $_"
-} finally {
-    Pop-Location
+Invoke-CleanUp $tempDir
```
Comment (twpayne) [inline assets/scripts/install.ps1:260]: Is this needed given the

```powershell
trap {
    Invoke-CleanUp $tempDir
    break
}
```

at line 189?

Comment (bradenhilton) [reply assets/scripts/install.ps1:260]: The trap will only execute after a terminating error. The "happy path" won't encounter a terminating error, so the trap will never execute, and it won't clean up the temporary archive file.

```console
❯ Get-Content .\test-trap.ps1
trap {
    Write-Host 'exited'
    break
}
Write-Host 'Hello world'

❯ .\test-trap.ps1
Hello world

❯ Get-Content .\test-trap-error.ps1
trap {
    Write-Host 'exited'
    break
}
throw 'Hello error'
Write-Host 'Hello world'

❯ .\test-trap-error.ps1
exited
Exception: C:\Users\Braden\test-trap-error.ps1:5
Line |
   5 |  throw 'Hello error'
     |  ~~~~~~~~~~~~~~~~~~~
     | Hello error
```

**Comment by bradenhilton** [comment]
This is strange.

I just tried alpine again while connected to a VPN on my host machine, and it worked:

```
❯ docker run -it alpine:latest

/ # apk update
...

/ # apk add powershell
...

/ # pwsh

PS /> iex "&{$(irm 'https://raw.githubusercontent.com/twpayne/chezmoi/e8f93fdf7daf6bc71fd2d4f2dcd1696f5ecd6786/assets/scripts/install.ps1')} -d"
DEBUG: BinDir: /bin
DEBUG: Tag: latest
DEBUG: ChezmoiArgs:
DEBUG: tempDir: /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc
DEBUG: goOS: linux
DEBUG: goArch: amd64
DEBUG: checking GitHub for tag latest
DEBUG: found tag v2.46.1 for latest
found version 2.46.1 for latest/linux/amd64
DEBUG: libcOutput: musl libc (x86_64)
Version 1.2.4_git20230717
Dynamic Program Loader
Usage: /lib/ld-musl-x86_64.so.1 [options] [--] pathname
DEBUG: binarySuffix:
DEBUG: archiveFormat: tar.gz
DEBUG: goOSExtra: -musl
DEBUG: archiveFilename: chezmoi_2.46.1_linux-musl_amd64.tar.gz
DEBUG: tempArchivePath: /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc/chezmoi_2.46.1_linux-musl_amd64.tar.gz
DEBUG: downloading https://github.com/twpayne/chezmoi/releases/download/v2.46.1/chezmoi_2.46.1_linux-musl_amd64.tar.gz
DEBUG: downloading https://github.com/twpayne/chezmoi/releases/download/v2.46.1/chezmoi_2.46.1_checksums.txt as string
DEBUG: extracting /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc/chezmoi_2.46.1_linux-musl_amd64.tar.gz to /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc
DEBUG: binaryFilename: chezmoi
DEBUG: tempBinaryPath: /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc/chezmoi
DEBUG: binary: /bin/chezmoi
installed /bin/chezmoi
DEBUG: removing /tmp/70842863-96fb-4f87-bb50-53f1076fa8fc
```

I disconnected from the VPN, deleted `/bin/chezmoi` and tried again:

```
PS /> iex "&{$(irm 'https://raw.githubusercontent.com/twpayne/chezmoi/e8f93fdf7daf6bc71fd2d4f2dcd1696f5ecd6786/assets/scripts/install.ps1')} -d"
DEBUG: BinDir: /bin
DEBUG: Tag: latest
DEBUG: ChezmoiArgs:
DEBUG: tempDir: /tmp/4c7672cb-eae2-4523-8377-56248ff00047
DEBUG: goOS: linux
DEBUG: goArch: amd64
DEBUG: checking GitHub for tag latest
DEBUG: removing /tmp/4c7672cb-eae2-4523-8377-56248ff00047
Write-Error:
Line |
 215 |  $realTag = Get-RealTag $Tag
     |             ~~~~~~~~~~~~~~~~
     | error retrieving GitHub release latest
```

It works for me on my host machine while not connected to the VPN:

```
❯ iex "&{$(irm 'https://raw.githubusercontent.com/twpayne/chezmoi/e8f93fdf7daf6bc71fd2d4f2dcd1696f5ecd6786/assets/scripts/install.ps1')} -d"
DEBUG: BinDir: C:\Users\Braden\bin
DEBUG: Tag: latest
DEBUG: ChezmoiArgs:
DEBUG: tempDir: C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107
DEBUG: goOS: windows
DEBUG: goArch: amd64
DEBUG: checking GitHub for tag latest
DEBUG: found tag v2.46.1 for latest
found version 2.46.1 for latest/windows/amd64
DEBUG: binarySuffix: .exe
DEBUG: archiveFormat: zip
DEBUG: goOSExtra:
DEBUG: archiveFilename: chezmoi_2.46.1_windows_amd64.zip
DEBUG: tempArchivePath: C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107\chezmoi_2.46.1_windows_amd64.zip
DEBUG: downloading https://github.com/twpayne/chezmoi/releases/download/v2.46.1/chezmoi_2.46.1_windows_amd64.zip
DEBUG: downloading https://github.com/twpayne/chezmoi/releases/download/v2.46.1/chezmoi_2.46.1_checksums.txt as string
DEBUG: extracting C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107\chezmoi_2.46.1_windows_amd64.zip to C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107
DEBUG: binaryFilename: chezmoi.exe
DEBUG: tempBinaryPath: C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107\chezmoi.exe
DEBUG: binary: C:\Users\Braden\bin\chezmoi.exe
installed C:\Users\Braden\bin\chezmoi.exe
DEBUG: removing C:\Users\Braden\AppData\Local\Temp\8e112f8b-7c17-4d61-aace-cadbced02107
```


I don't know if I can be bothered diagnosing that problem.

I'm happy to merge if you're happy with my explanation in https://github.com/twpayne/chezmoi/pull/3569#discussion_r1496527694.

File: assets/scripts/install.ps1 (hunk 3 above)
Comment (twpayne) [reply assets/scripts/install.ps1:260]: Thank you for the explanation :)

**Review by twpayne — APPROVED** [review]
> I'm happy to merge if you're happy with my explanation in [#3569 (comment)](https://github.com/twpayne/chezmoi/pull/3569#discussion_r1496527694).

Yes, absolutely. The problem is clearly not related this script.

Please merge - this is a brilliant contribution!

**Comment by bradenhilton** [comment]
For info, I set Docker to use Google DNS, and it now works without using a VPN, so I think the issue is an interaction between Docker and my ISP's DNS.

**Pull request merged**